object_intern.go

```diff
@@ -100,49 +103,88 @@ func (oi *ObjectIntern) DecompressString(in string) (string, error) {
 // If the object is found in the store its reference count is increased by 1.
 // If the object is added to the store its reference count is set to 1.
 func (oi *ObjectIntern) AddOrGet(obj []byte, safe bool) (uintptr, error) {
-	var addr gos.ObjAddr
-	var ok bool
-	var err error
-	var objString string
 
-	objComp := obj
+	// if either of these two terms is true then the rest of this block
+	// requires a lot of allocations
+	if (oi.conf.Compression != None) || (safe && oi.conf.Compression == None) {
+		objComp := obj
 
-	if oi.conf.Compression != None {
-		objComp = oi.compress(obj)
-	}
+		if oi.conf.Compression != None {
+			// this returns a new byte slice, so we don't need to check for safe
+			objComp = oi.compress(obj)
+		}
 
-	// the only case we need to handle specially is when compression
-	// is turned off and the user has requested that the operation be safe
-	if safe && oi.conf.Compression == None {
-		// create a copy so we don't modify the original []byte
-		objComp = make([]byte, len(obj))
-		copy(objComp, obj)
-	}
+		// the only case we need to handle specially is when compression
+		// is turned off and the user has requested that the operation be safe
+		if safe && oi.conf.Compression == None {
+			// create a copy so we don't modify the original []byte
+			objComp = make([]byte, len(obj))
+			copy(objComp, obj)
+		}
 
-	// string used for the index
-	objString = string(objComp)
+		// acquire lock
+		oi.Lock()
+
+		// try to find the object in the index
+		addr, ok := oi.objIndex[string(objComp)]
+		if ok {
+			// increment reference count by 1
+			(*(*uint32)(unsafe.Pointer(addr + uintptr(len(objComp)))))++
+			oi.Unlock()
+			return addr, nil
+		}
+
+		// The object is not in the index therefore it is not in the store.
+		// We need to set its initial reference count to 1 before adding it.
+		//
+		// The object store backend has no knowledge of a reference count, so
+		// we need to manage it at this layer. Here we add 4 bytes to be used
+		// henceforth as the reference count for this object. Reference count is
+		// always placed as the LAST 4 bytes of an object and is NEVER compressed.
+		objComp = append(objComp, InitialRefCount...)
+		addr, err := oi.store.Add(objComp)
+		if err != nil {
+			oi.Unlock()
+			return 0, err
+		}
+
+		// remove ref count from end before converting to string
+		objString := string(objComp[:len(objComp)-4])
+
+		// set objString data to the object inside the object store
+		((*reflect.StringHeader)(unsafe.Pointer(&objString))).Data = addr
+
+		// add the object to the index
+		oi.objIndex[objString] = addr
+
+		oi.Unlock()
+		return addr, nil
+	}
 
+	// if neither of those terms is true then we can avoid costly allocations
 	// acquire lock
 	oi.Lock()
```

Comment:
Assuming that in most cases when this method is called, the object is already present, shouldn't we use an `RLock()` until we know that we have to insert the object? same for the above compressed/save insert

Reply:
Good point, i need to review all the locks anyways